### PR TITLE
Do not add barrier after send

### PR DESCRIPTION
**Pull request by whchung**
On AMD GCN ISA, the semantic of nontemporal store already guarantee data fully visible on remote GPU. There shouldn't be no barriers needed after send, under all-to-all communication pattern.

Empirical performance benchmarking shows improvement in smaller message sizes with correct results.

### CODE REVIEW

**Comment by yzygitzh** [comment]
Hi,

This barrier was designed to avoid data corruption when data from the same address is touched by different threads.
Imagine there is 1 thread block which has 2 threads, and they're doing send first and then recv.
There are 4 data slots (0, 1, 2, 3).
For send, thread 0 is going to send (0, 1) and thread 1 is going to send (2, 3).
For recv, thread 0 is going to recv (0) and thread 1 is going to recv (1).
Now, without barrier, thread 1 might go ahead of thread 0 and wrongly overwrites data slot 1.

This is not a common case while some special algorithms we use will have different send size and recv size. One potential way to solve this is to differentiate "send w/ barrier" and "send w/o barrier" in MSCCL algorithm interface.

File: src/collectives/device/prims_ll.h [hunk 1]
```diff
@@ -744,8 +744,6 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL, P2p>:
   // MSCCL primitives
   __device__ void sendWithBarrier(intptr_t inpIx, int eltN) {
     send(inpIx, eltN);
-    // This is the only primitive.instruction where there is no barrier at the end, add it
-    barrier();
```
Comment (nusislam) [inline src/collectives/device/prims_ll.h:748]: Will removing this barrier cause any synchronization issue?

**Comment by whchung** [comment]
Closing this PR after yielding no perceivable improvements on target machines.

**Comment by whchung** [comment]
Reviving the PR as it proves to be helpful on upcoming SKUs. The logic needs to modified so it's only applied under LL protocl though.

**Comment by yzygitzh** [comment]
> Reviving the PR as it proves to be helpful on upcoming SKUs. The logic needs to modified so it's only applied under LL protocl though.

Let me help drive a PR to introduce "swb" (send with barrier) instruction in XML to enable this while not breaking non-safe algorithms.

**Review by wenkaidu — APPROVED** [review]

**Pull request merged**